### ISSUE TITLE
Uncaught TypeError: Cannot read property 'filename' of null

### ISSUE THREAD

**Issue by leoliu** [pqina/filepond#110]
### Summary

I had something like the following to remove a file after upload succeeded.

```javascript
    this.pond.on("processfile", (error, file) => {
      error
        ? toast({ type: "error", text: error.code + " " + error.body })
        : this.uploadSucceeded(file);
    });

  uploadSucceeded = file => {
    if (!file) return;
    this.pond.removeFile(file);
  };

```

 However I am seeing a crash in production: Uncaught TypeError: Cannot read property 'filename' of null at Object.DID_COMPLETE_ITEM_PROCESSING......

I traced the crash to this code in filepond.js

```javascript
  var itemProcessed = function itemProcessed(_ref4) {
    var root = _ref4.root,
      action = _ref4.action;

    // will also notify the user when FilePond is not being used, as the user might be occupied with other activities while uploading a file

    var item = root.query('GET_ITEM', action.id);
    var filename = item.filename;
    var label = root.query('GET_LABEL_FILE_PROCESSING_COMPLETE');

    assist(root, filename + ' ' + label);
  };

```

UA:

"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/66.0.3359.117 Safari/537.36"

### Additional information

| Environment  | Version
|------------- | -----------
| OS           | MacOS / Windows / iOS / Linux / ...
| Device       | iPhone SE / Galaxy Note 9 / ...
| Browser      | Safari 11.1.1 / Edge 15 / ...

**Comment by rikschennink** [comment]
What version of FilePond are you running?

**Comment by rikschennink** [comment]
Hmm odd, the events should be async, what is happening is that you're removing the file before the process file thread has had time to complete (which of course should not happen). Wrapping the `uploadSucceeded` call in a `setTimeout` should fix this for now.

```js
setTimeout(() => this.uploadSucceeded(file), 0);
```

**Comment by leoliu** [comment]
Thanks for the workaround. I'll try it and report back.

I also have a button where users click to upload all files (instantUpload is false). The button click triggers the following function. Could this be the problem (some race issue maybe)?

```javascript
  async submit(e) {
    e.preventDefault();
    this.files.forEach(file => {
      if (
        [
          FilePond.FileStatus.IDLE,
          FilePond.FileStatus.PROCESSING_ERROR
        ].includes(file.status)
      ) {
        this.addMetadata(file);
        this.pond.processFile(file).then(this.uploadSucceeded);
      }
    });
  }
```

**Comment by leoliu** [comment]
The workaround doesn't work. I am still seeing the same crash.

**Comment by rikschennink** [comment]
Alright that’s to bad, will take a look at this on Monday and get back to you.

**Comment by leoliu** [comment]
Sounds good ;) look forward to the fix.

**Comment by leoliu** [comment]
I found a fix. In the `submit` function above change `this.pond.processFile(file).then(this.uploadSucceeded);` to `this.pond.processFile(file);`.

Looks like some code in filepond doesn't expect the `file` to be removed.

**Comment by rikschennink** [comment]
Just published version `2.1.1` which should fix this issue, it adds a tiny delay to the resolve call of `processFile`, not the best solution but it seems to work. Please confirm it resolves the issue.

**Comment by leoliu** [comment]
Yes, it fixes the issue. Thanks.
